src/opossum_lib/file_generation.py

```diff
@@ -1,89 +1,159 @@
             raise RuntimeError(
                 "A tree should always have a node without incoming edge."
             )
-
-        opossum_information["resources"][source] = tree_to_dict(
-            connected_subgraph, source
-        )
-        opossum_information["attributionBreakpoints"].extend(
-            get_breakpoints(connected_subgraph, source)
+        resources_list = []
+        for node in connected_subgraph.nodes():
+            path: List[str] = shortest_path(connected_subgraph, source, node)
+            path_with_labels: List[str] = _replace_node_ids_with_labels(
+                path, connected_subgraph
+            )
+            file_path: str = _create_file_path_from_graph_path(connected_subgraph, path)
+
+            resources_list.append(_create_nested_dict(path_with_labels))
+            if _node_represents_a_spdx_element(connected_subgraph, node):
+                create_attribution_and_link_with_resource(
+                    external_attributions,
+                    resources_to_attributions,
+                    file_path,
+                    node,
+                    connected_subgraph,
+                )
+
+            else:
+                attribution_breakpoints.append(file_path)
+
+        resources.update(
+            reduce(
+                lambda dict1, dict2: _merge_nested_dicts(dict1, dict2), resources_list  # type: ignore  # noqa
+            )
         )
-
+    opossum_information = OpossumInformation(
+        metadata=metadata,
+        resources=resources,
+        externalAttributions=external_attributions,
+        resourcesToAttributions=resources_to_attributions,
+        attributionBreakpoints=attribution_breakpoints,
+    )
     return opossum_information
 
 
-def tree_to_dict(tree: DiGraph, source: str) -> Union[int, Dict[str, Any]]:
-    successors = list(tree.successors(source))
-    if not successors:
-        return 1
-    branch_dict = {}
-    for successor in successors:
-        successor_name = tree.nodes[successor]["label"]
-        branch_dict[successor_name] = tree_to_dict(tree, successor)
-    return branch_dict
-
-
-def get_breakpoints(tree: DiGraph, source: str) -> List[str]:
-    breakpoints = []
-    for node in tree.nodes():
-        if not _node_represents_a_spdx_element(tree, node):
-            path = shortest_path(tree, source, node)
-            breakpoints.append(_create_file_path_from_graph_path(tree, path))
+def create_attribution_and_link_with_resource(
+    external_attributions: Dict[str, OpossumPackage],
+    resources_to_attributions: Dict[str, List[str]],
+    file_path: str,
+    node: str,
+    tree: DiGraph,
+) -> None:
+    if isinstance(tree.nodes[node]["element"], Package):
+        external_attributions.update(
+            create_package_attribution(tree.nodes[node]["element"])
+        )
+        resources_to_attributions[file_path] = [
+            tree.nodes[node]["element"].spdx_id,
+            "package-identifier",
```

Comment:
What is the idea behind this "package-identifier"? We are linking two Opossum Packages to the resource?

Reply:
The idea is to somehow identify the element as a package (snippet, file resp.). So by adding this generic "identifier" attribution we keep the information which kind of spdx element the resource is.